File: 0127-word-ladder/0127-word-ladder.py
class Solution:
  def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
    if endWord not in wordList: return 0
    
    patterns = collections.defaultdict(list)
    def getPatterns(word: str) -> None:
      for i in range(len(word)):
        key = word[:i]+"*"+word[i+1:]
        patterns[key].append(word)
    for word in [beginWord]+wordList:
      getPatterns(word)
    
    pathQ, visited, currPath, currNode = [[beginWord]], set([beginWord]), None, None
    while pathQ:
      currPath = pathQ.pop(0)
      currNode = currPath[-1]
      if currNode == endWord: return len(currPath)
      for i in range(len(currNode)):
        key = currNode[:i]+"*"+currNode[i+1:]
        for nb in patterns[key]:
          if nb != currNode and nb not in visited:
            visited.add(nb)
            pathQ.append(currPath+[nb])
    return 0

File: 0127-word-ladder/0127-word-ladder.rb
# @param {String} begin_word
# @param {String} end_word
# @param {String[]} word_list
# @return {Integer}
def get_patterns(word, patterns)
  (0...word.length).each do |i|
    key = word[0...i]+"*"+word[i+1..-1]
    patterns[key] << word
  end
  nil
end

def ladder_length(begin_word, end_word, word_list)
  return 0 if !word_list.include?(end_word)
  
  patterns = Hash.new{ |h, k| h[k] = Array.new }
  [begin_word]+word_list.each{ |word| get_patterns(word, patterns) }
  path_q, visited, curr_path, curr_node = [[begin_word]], [begin_word].to_set, nil, nil
  while !path_q.empty?
    curr_path = path_q.shift
    curr_node = curr_path[-1]
    return curr_path.length if curr_node == end_word
    (0...curr_node.length).each do |i|
      key = curr_node[0...i]+"*"+curr_node[i+1..-1]
      patterns[key].each do |nb|
        if nb != curr_node && !visited.include?(nb)
          visited << nb
          path_q << curr_path+[nb]
        end
      end
    end
  end
  0
end